modules/packages/SharedObject.chpl

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright 2004-2017 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+
+   :record:`Shared` (along with :record:`Owned`) manage the deallocation
+   of a class instance. :record:`Shared` is meant to be used when many
+   different references will exist to the object and these references
+   need to keep the object alive.
+
+   To use :record:`Shared`, allocate a class instance following this
+   pattern:
+
+   .. code-block:: chapel
+
+     var mySharedObject = new Shared(new MyClass(...));
+
+   now, when mySharedObject and any copies of it go out of scope, the class
+   instance it refers to will be deleted.
+
+   Now, copy initializing or assigning from mySharedObject will make
+   other variables refer to the same class instance. The class instance
+   will be deleted after all of these references go out of scope.
+
+   .. code-block:: chapel
+
+     var globalSharedObject:Shared(MyClass);
+
+     proc makeGlobalSharedObject() {
+       var mySharedObject = new Shared(new MyClass(...));
+       globalSharedObject = mySharedObject;
+       // now mySharedObject is deinitialized, but the MyClass
+       // instance is not deleted until globalSharedObject is deinitialized.
+     }
+
+ */
+module SharedObject {
+
+  pragma "no doc"
+  class ReferenceCount {
+    var count: atomic int;
+
+    // count should be initialized to 1 in default initializer.
+    // Pretty much any strategy can do that.
+    // There is the wrinkle that initializing an atomic
+    // currently happens with a.write(1). We could presumably
+    // support = for initialing atomics, however.
+    proc ReferenceCount() {
+      //count = 1;      // this should work
+      //super.init();
+      count.write(1); // how you'd write it now
+    }
+
+    proc retain() {
+      count.add(1);
+    }
+    proc release() {
+      var oldValue = count.fetchSub(1);
+      return oldValue - 1;
+    }
+  }
+
+
+  /*
+
+     The :record:`Shared` manages the deletion of a class instance in a way
+     that supports multiple owners of the class instance.
+
+     This is currently implemented with task-safe reference counting.
+
+   */
+  record Shared {
+    pragma "no doc"
+    var p;                 // contained pointer (class type)
+    pragma "no doc"
+    var pn:ReferenceCount; // reference counter
```

Comment:
I suggest making both `p` and `pn` be `private`.
We don't want users to access them, right?

Correspondingly, change tests to use `.borrow()` instead of `.p`.
We should be testing as if we are users, right? :)